xapian-applications/omega/handler_libarchive.cc

```diff
@@ -0,0 +1,122 @@
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "metaxmlparse.h"
+#include "opendocparse.h"
+#include "str.h"
+
+#include<archive.h>
+#include<archive_entry.h>
+
+#include <cstring>
+
+using namespace std;
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    try {
+	struct archive* archive_obj;
+	struct archive_entry* entry;
+	int status_code;
+	const char* file = filename.c_str();
+	archive_obj = archive_read_new();
+	archive_read_support_filter_all(archive_obj);
+	archive_read_support_format_all(archive_obj);
+	status_code = archive_read_open_filename(archive_obj, file, 10240);
+
+	if (status_code != ARCHIVE_OK) {
+	    error = "Libarchive failed to open the file " + filename;
+		return false;
+	}
+
+	size_t total;
+	ssize_t size;
+	string s;
+
+	// extracting data from content.xml and styles.xml
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    if (strcmp(archive_entry_pathname(entry), "content.xml") == 0) {
+		total = archive_entry_size(entry);
+		char buf1[total];
+		size = archive_read_data(archive_obj, buf1, total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data from "
+		    "content.xml";
+		    return false;
+		}
+
+		s += str(buf1);
+		} else if (strcmp(
+			archive_entry_pathname(entry), "styles.xml") == 0) {
+		total = archive_entry_size(entry);
+		char buf2[total];
+		size = archive_read_data(archive_obj, buf2, total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data from "
+		    "styles.xml";
+		    return false;
+		}
+
+		s += str(buf2);
+		OpenDocParser parser;
+		parser.parse(s);
+		dump = parser.dump;
+		} else if (strcmp(
+			archive_entry_pathname(entry), "meta.xml") == 0) {
+		total = archive_entry_size(entry);
+		char buf3[total];
+		size = archive_read_data(archive_obj, buf3, total);
+
+		if (size <= 0) {
+		    // error handling
+		}
+
+		s = str(buf3);
+		MetaXmlParser metaxmlparser;
+		metaxmlparser.parse(s);
+		title = metaxmlparser.title;
+		keywords = metaxmlparser.keywords;
+		author = metaxmlparser.author;
+		}
+	}
+	status_code = archive_read_free(archive_obj);
+	if (status_code != ARCHIVE_OK) {
+	    return false;
```

Comment:
Not sure if we should return `false` here and drop all the extracted information.

Reply:
It seems that if `archive_read_free` fails, some file descriptors can remain opened and that can cause problems(run out of file descriptors) in some scenarios. 
Have @Exter-dg  seen [`archive_error_string`](https://nxmnpg.lemoda.net/3/archive_error_string)? Maybe we can get more information to append to the error string.

Reply:
Yes, I read about 'archive_error_string`, it can provide us with more information about the error produced incase archive_read_free() fails. I have changed this.